libcxx/include/__type_traits/is_trivially_relocatable.h

```diff
@@ -0,0 +1,51 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___TYPE_TRAITS_IS_TRIVIALLY_RELOCATABLE_H
+#define _LIBCPP___TYPE_TRAITS_IS_TRIVIALLY_RELOCATABLE_H
+
+#include <__config>
+#include <__type_traits/is_trivially_destructible.h>
+#include <__type_traits/is_trivially_move_constructible.h>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+//------------------------------------------------------WARNING-------------------------------------------------------//
+//
+// This type trait should only be used where ABI stability isn't relevant or can be assured by other means,
+// like in function bodies or where the difference can be detected at runtime. If this trait is used in ABI-sensitive
+// environments it has to be ensured that Clang and GCC can interpret the data produced by each other. If you are not
+// certain that this is the case DO NOT USE THIS TYPE TRAIT
+//
+//--------------------------------------------------------------------------------------------------------------------//
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if __has_builtin(__is_trivially_relocatable)
+
+template <class _Type>
+struct __libcpp_is_trivially_relocatable : integral_constant<bool, __is_trivially_relocatable(_Type)> {};
+
+#else
+
+template <class _Type>
+using __libcpp_is_trivially_relocatable =
+    integral_constant<bool, is_trivially_move_constructible<_Type>::value && is_trivially_destructible_v<_Type>::value>;
+
+#endif // __has_builtin(__is_trivially_relocatable)
+
+#if _LIBCPP_STD_VER > 14
+template <class _Type>
+constexpr bool __libcpp_is_trivially_relocatable_v = __libcpp_is_trivially_relocatable<_Type>::value;
```

Comment:
```suggestion
inline constexpr bool __libcpp_is_trivially_relocatable_v = __libcpp_is_trivially_relocatable<_Type>::value;
```
Nit: I question the need for this variable template at all.

Reply:
Does adding `inline` change anything in reality or is it just that it's technically correct?
I'd rather keep this, since it's trivial to replace it with `__libcpp_is_trivially_relocatable_v = __is_trivially_relocatable(_Type)` once we get the attribute. This is currently only instantiating `__libcpp_is_trivially_relocatable` because I want to specialize it for some types later.

Reply:
> Does adding `inline` change anything in reality or is it just that it's technically correct?

Not coincidentally, I have a blog post on this. :) https://quuxplusone.github.io/blog/2022/07/08/inline-constexpr/

> I'd rather keep this [variable template instead of using `::value`], since it's trivial to replace it with `__libcpp_is_trivially_relocatable_v = __is_trivially_relocatable(_Type)` once we get the attribute.

FYI, Clang already has the `__is_trivially_relocatable` builtin, so you could already do that; I mean there's no need for the class template _already_, if you're actually _not_ trying to sneak in a full implementation of a P1144-compatible (STL-style) type trait.

However, I have a counterproposal: To avoid unnecessary merge conflicts and [error-prone partial specializations](https://www.youtube.com/watch?v=SGdfPextuAU&ab_channel=CppNow&t=24m50s), I think you should take one of the following approaches to disentangle `move_only_function` from `is_trivially_relocatable`:
- Take #2 as a prerequisite to this one. It adds a trunk-compatible `__libcpp_is_trivially_relocatable` type trait with minimal diffs versus my branch. Then your `move_only_function` patch can simply use that trait. (That is, "trivially relocatable first, function second.")
- Eliminate "trivially relocatable" concerns from `move_only_function` entirely, for now. Change this line to use `if constexpr (is_trivially_copyable_v<T>)` instead. This won't lose you anything in practice. And then if someone (e.g. you) in the future wants to add `__libcpp_is_trivially_relocatable` in a separate patch, they can just update the `if constexpr` line at the same time. (That is, "function first, trivially relocatable second.")